### ISSUE TITLE
http client sometimes hangs when receiving a body in excess of response-buffer-size

### ISSUE THREAD

**Issue by pbwolf** [ztellman/aleph#193]
The enclosed small web client uses aleph.http ("0.4.1-beta3") to retrieve a byte array with `aleph.http/get`.  Sometimes, the web response (the Deferred) never materializes!  The program sits, seemingly forever, after the "Sending..." message and before the "Received".

I have only seen the problem arise when the byte array would be bigger than response-buffer-size and the client and server are running on the same computer.  Even then, it is hit-or-miss.

Here is the web client program:

```
(ns giblet.podunk
  (:require [aleph.http :as http]))
(defn -main
  [url]
  (let [pool (http/connection-pool
              {:connections-per-host 1
               :connection-options {:keep-alive? true}})]
      (dotimes [n 42000000]
        (println "Sending request" n)
        (let [hr (deref
                  (http/get url
                            {:pool pool
                             :as :byte-array
                             :query-params {"n" (str n)}}))]
          (println "Received" n (alength (:body hr)))))))
```

Adding the connection-option `{:response-buffer-size bigger-number}`, to accommodate the biggest-imaginable response body, seems to work around the problem.  But doing so probably shouldn't be necessary, since that option is documented as "This does *not* represent the maximum size response that the client can handle (which is unbounded), and is only a means of maximizing performance".

**Comment by ztellman** [comment]
If you don't consume the body, the response won't be considered complete, and subsequent requests won't be able to be made.  By increasing the size of the response buffer, you're working around this by making the consumption implicit, but I would consider this behavior "expected".  I'm going to mark this closed, but let me know if you think this is incorrect.

**Comment by pbwolf** [comment]
Aleph dispenses the body as a byte array, so I think it's consumed.  Moreover, when the program runs distantly from the server, it does not seem to run into the problem.

**Comment by ztellman** [comment]
Oh, I missed the `:as` clause.  This may be an issue where the response comes fast enough that the `bs/to-byte-array` call occurs on the Netty thread, rather than the response handling pool, which is meant to prevent these sorts of deadlocks.

Reopening, I'll take a closer look.  Sorry for the confusion on my end.

**Comment by ztellman** [comment]
Okay, this shouldn't have been sitting here all this time, my apologies.  I'm fairly sure this has been fixed by the changes in `[byte-streams "0.2.1-alpha3"]`, where I discovered a potential deadlock when transforming Manifold streams to InputStreams.  I'm going to do some tests on my end, but if you can confirm on your end that would be greatly appreciated.

**Comment by pbwolf** [comment]
I overrode Aleph's byte-streams dependency with 0.2.1-alpha3 and also overrode byte-streams 0.2.1-alpha3's dependency on manifold (it needs consume-async).  But the program still hung.

**Comment by ztellman** [comment]
Hmm, okay.  I'll take another look.

**Comment by ztellman** [comment]
I'm having trouble reproducing this.  A few questions:

* where was the server located?  If on another machine, what was the RTT between the servers?
* how big were the responses, at their largest?

**Comment by pbwolf** [comment]
I most often see failures retrieving a resource larger than 64k.  For instance, the
program hung on iteration 5, 55, 23, 43, and 106 in five consecutive
runs when retrieving a resource of 66931 bytes with a buffer size of 16384.  
With a buffer size of 65536 it hung on iteration 255, 236, 244, 407, and 251.  
I invoked the program with java -cp (uberjar) -m clojure.main giblet.podunk 
using Java's default options.  I have tried with the server on the same computer 
and on the local area network.   I don't know about RTT.

**Comment by ztellman** [comment]
That's strange, I tried to reproduce it with this: https://github.com/ztellman/aleph/blob/master/test/aleph/http_test.clj#L290-L300 and could after tens of thousands of iterations.

I guess I'll just run your code exactly and see if that changes anything.

**Comment by ztellman** [comment]
Still unable to reproduce it with this:

```clj
(ns alephtest.core
  (:require [aleph.http :as http])
  (:gen-class))


(defn -main
  [& args]

  (let [body (byte-array 1e5)]
    (http/start-server
      (fn [req]
        {:status 200
         :body body})
      {:port 8888}))

  (let [pool (http/connection-pool
               {:connections-per-host 1
                :connection-options {:keep-alive? true}})]
    (dotimes [n 42000000]
      (println "Sending request" n)
      (let [hr (deref
                 (http/get "http://localhost:8888"
                   {:pool pool
                    :as :byte-array
                    :query-params {"n" (str n)}}))]
        (println "Received" n (alength (:body hr)))))))
```

Is there something appreciably different between my server and yours?

**Comment by pbwolf** [comment]
alephtest.core appears to work fine.  I also tried Apache serving a file (no hangs).  I tried reverse-proxying a problematic server with Apache, and the Aleph client still hung, but less frequently.  Perhaps there is a difference of timing or chunk size compared to alephtest.core.  (Using curl to fetch the problematic server's resource, I see it says "Transfer-Encoding: chunked".)

**Comment by ztellman** [comment]
Okay, modified test case with delayed chunks:

```clj
(ns alephtest.core
  (:require
    [aleph.http :as http]
    [manifold.stream :as s])
  (:gen-class))

(defn -main
  [& args]

  (http/start-server
    (fn [req]
      {:status 200
       :body (->> (s/periodically 0.1 #(byte-array 1024))
               (s/transform (take 1e2)))})
    {:port 8888})

  (let [pool (http/connection-pool
               {:connections-per-host 1
                :connection-options {:keep-alive? true}})]
    (dotimes [n 42000000]
      (println "Sending request" n)
      (let [hr (deref
                 (http/get "http://localhost:8888"
                   {:pool pool
                    :as :byte-array
                    :query-params {"n" (str n)}}))]
        (println "Received" n (alength (:body hr)))))))
```

I still can't make this fail.  Maybe you can give me a tcpdump of the traffic for the problematic server?  It would look something like:

```
sudo tcpdump -i any src port PORT or dst port PORT -c 1000000 -w FILE
```

**Comment by pbwolf** [comment]
tcpdump captured three hangs.  It did not emit any time stamps (I also tried -ttt).  Observations:

 - The chunk sizes are consistently (hex) 8000, 2000, 2000, 2000, ..., remainder, 0.

 - A final chunk (0) is the last captured packet.

**Comment by ztellman** [comment]
By "final", you mean the last packet that's sent before the hang?

**Comment by ztellman** [comment]
Can you share the gzipped dump somewhere for me to download?  I'm trying to mimic the chunk size, too, and not having any luck.  Maybe it would help to just compare the dumps of my test case and yours.

**Comment by ztellman** [comment]
Yeah, there was a bug in `[byte-streams 0.2.1-alpha3]`, try `alpha4` instead.

**Comment by pbwolf** [comment]
OK, I updated to [byte-streams "0.2.1-alpha4"].  There might be something going on with the stream the response gets copied to.

I put a println into client-handler in aleph.http.client:

                 ;; already have a stream going
                 (do
                   (println (.getId (Thread/currentThread)) "HttpContent (non-last)")

and observed that in an iteration of giblet.podunk that hangs, with
response buffer size 16384 (the unluckiest buffer size I chanced to
find), the hang usually comes after the 2nd of those printlns. 
Not, in other words, near the end of the response, which tcpdump seems
to show has been entirely delivered and acknowledged.  But other
response buffer sizes that I tried, did not hang anymore, during a brief test. 

Then I put (Thread/sleep 1) after the println (i.e., before
netty/put!), and I have not seen a hang since, even with response
buffer size 16384.

**Comment by ztellman** [comment]
Right, this is almost certainly a case of backpressure being established and not being undone once the body starts to be consumed.  The TCP stack has received the data, but it's queued up in Netty or somewhere upstream.

This is helpful, thank you.  I'll look at it some more.

**Comment by ztellman** [comment]
I'm taking a flyer here, but can you try it again with `[aleph 0.4.1-beta5]` without any other dependencies?

**Comment by pbwolf** [comment]
It still hangs.

**Comment by ztellman** [comment]
Okay, I don't think I'm able to diagnose this without the ability to reproduce it locally.  What information can you give me about the web server you're using?

**Comment by pbwolf** [comment]
Here's a modified amalgam of the above little programs, incorporating both server and client as you illustrated, that I have tried on 2 computers.  It hangs within 20,000 iterations on both of them.  It hangs with or without the Thread/sleep, but it hangs faster with the sleep than without.

openjdk version "1.8.0_72"
OpenJDK Runtime Environment (build 1.8.0_72-b15)
OpenJDK 64-Bit Server VM (build 25.72-b15, mixed mode)

--- 8< ---
```
(defproject giblet/podunk "0.1.0-SNAPSHOT"
  :dependencies [[org.clojure/clojure "1.8.0"]
                 [aleph "0.4.1-beta5"]]
  :main giblet.podunk)
```
--- 8< ---
```
(ns giblet.podunk
  (:require [aleph.http :as http]
            [manifold.stream :as s]))
(defn -main
  [port response-buffer-size]

  ;; server
  (let [small-chunk  (byte-array 0x1234)
        medium-chunk (byte-array 0x2000)
        large-chunk  (byte-array 0x8000)]
    (http/start-server
     (fn [req]
       (let [s (s/stream)]
         (s/put! s large-chunk) 
         (future
           (Thread/sleep 1)
           (s/put! s medium-chunk)
           (s/put! s medium-chunk)
           (s/put! s medium-chunk)
           (s/put! s medium-chunk)
           (s/put! s small-chunk)
           (s/close! s))
         {:status 200
          :body s}))
     {:port (Integer/parseInt port)}))
  
  ;; client
  (let [url (str "http://localhost:" port "/")
        pool (http/connection-pool
              {:connections-per-host 1
               :connection-options {:keep-alive? true
                                    :response-buffer-size
                                    (Integer/parseInt response-buffer-size)}})]
    (dotimes [n 42e6]
      (println "Sending request" n)
      (let [hr (deref
                (http/get url
                          {:pool pool
                           :as :byte-array
                           :query-params {"n" (str n)}}))]
        (println "\tReceived" n (alength (:body hr)))))))
```
--- 8< ---

Either
```
 lein run -m giblet.podunk 8088 16384
```
Or
```
 lein clean

 lein uberjar

 java -cp target/podunk-0.1.0-SNAPSHOT-standalone.jar clojure.main -m giblet.podunk 8088 16384
```

**Comment by ztellman** [comment]
Thank you, I really appreciate your ongoing efforts to help me figure this out.  I'll see what I can find.

**Comment by ztellman** [comment]
I think I've finally figured it out!  Adding `[manifold 0.1.3-alpha1]` to the deps seems to prevent the issue from happening.

**Comment by pbwolf** [comment]
With manifold 0.1.3-alpha1, the test often runs without hanging if I sit on my hands.  But I could sometimes provoke a hang by opening or closing another terminal window -- I guess the foreground change induces a task switch that shakes up the JVM thread scheduling rhythm.  Here is a variant of the test that uses 20 client threads.  On my 4-thread laptop, most of the threads get stuck before their 10000th iteration. 

```
(defproject giblet/podunk "0.1.0-SNAPSHOT"
  :dependencies [[org.clojure/clojure "1.8.0"]
                 [aleph "0.4.1-beta5"] 
                 [manifold "0.1.3-alpha1"]]
  :main giblet.podunk)
```

```
(ns giblet.podunk
  (:require [aleph.http :as http]
            [manifold.stream :as s]))
(defn -main
  [port response-buffer-size]

  ;; server
  (let [small-chunk  (byte-array 0x1234)
        medium-chunk (byte-array 0x2000)
        large-chunk  (byte-array 0x8000)]
    (http/start-server
     (fn [req]
       (let [s (s/stream)]
         (s/put! s large-chunk) 
         (future
           (Thread/sleep 1)
           (s/put! s medium-chunk)
           (s/put! s medium-chunk)
           (s/put! s medium-chunk)
           (s/put! s medium-chunk)
           (s/put! s small-chunk)
           (s/close! s))
         {:status 200
          :body s}))
     {:port (Integer/parseInt port)}))
  
  ;; client
  (let [client
        (fn [label pause]
          (let [url (str "http://localhost:" port "/")
                pool (http/connection-pool
                      {:connections-per-host 1
                       :connection-options {:keep-alive? true
                                            :response-buffer-size
                                            (Integer/parseInt response-buffer-size)}})]
            (try
              (dotimes [n 42e6]
                (println label "Sending request" n)
                (let [hr (deref
                          (http/get url
                                    {:pool pool
                                     :as :byte-array}))]
                  (println label "\tReceived" n (alength (:body hr)))
                  (pause)))
              (catch Exception e
                (println label "\tthrew" e)
                (System/exit 2)))))]
    (let [n-clients 20]
      (dotimes [i n-clients]
        (let [label (apply str (assoc
                                (into [] (repeat n-clients " "))
                                i
                                (char (+ (long \A) i))))]
          (future 
                  (client label #(Thread/sleep 0 (* i 5000)))))))))
```
Either
```
lein clean && lein uberjar
java -cp target/podunk-0.1.0-SNAPSHOT-standalone.jar clojure.main -m giblet.podunk 8088 16384
```
or
```
lein clean
lein run -m giblet.podunk 8088 16384
```

**Comment by ztellman** [comment]
Okay, yesterday's fix was a little voodoo, but now I've finally realized where the race condition is.  I am almost certain `0.1.3-alpha2` fixes your problem.  Again, very grateful for your patience with all of this.

**Comment by pbwolf** [comment]
Hot dog!  It runs happily with manifold 0.1.3-alpha2.  The full combination of libraries is:

```
[org.clojure/clojure "1.8.0"]
[aleph "0.4.1-beta5"] 
[manifold "0.1.3-alpha2"]
```

**Comment by ztellman** [comment]
Very happy to hear that.  I made a few changes along the way, but the core issue was https://github.com/ztellman/manifold/blob/master/src/manifold/stream.clj#L912-L921, where I previously didn't have the `locking` scope.  Some interleavings of the atomic operations on `buffer-size` and `last-put` would cause issues, and it shouldn't have taken me this long to figure that out.

Again, thank you for your efforts in helping me track this down, and your patience along the way.  I was waiting on doing a full release until I figured this out, so it's a big weight off my shoulders.
